nixos/tests/pleroma.nix

```diff
@@ -0,0 +1,270 @@
+    - manually upload your pleroma secrets to /var/lib/pleroma/secrets.exs
+    - use a deployment tool such as morph or NixOps to deploy your secrets.
+    */
+  provision-secrets = pkgs: pkgs.writeScriptBin "provision-secrets" ''
+    set -eux
+    cp "${pleroma-conf-secret pkgs}" "/var/lib/pleroma/secrets.exs"
+    chown pleroma:pleroma /var/lib/pleroma/secrets.exs
+  '';
+
+  /* For this NixOS test, we *had* to store this secret to the store.
+    Keep in mind the store is world-readable, it's the worst place
+    possible to store *any* secret. **DO NOT DO THIS IN A REAL WORLD
+    DEPLOYMENT**.
+  */
+  provision-user = pkgs: pkgs.writeScriptBin "provision-user" ''
+    set -eux
+
+    # Waiting for pleroma to be up.
+    timeout 5m bash -c 'while [[ "$(curl -s -o /dev/null -w '%{http_code}' https://pleroma.nixos.test/api/v1/instance)" != "200" ]]; do sleep 2; done'
+    pleroma_ctl user new jamy jamy@nixos.test --password 'jamy-password' --moderator --admin -y
+  '';
+
+  tls-cert = pkgs: pkgs.runCommandNoCC "selfSignedCerts" { buildInputs = [ pkgs.openssl ]; } ''
+    openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -nodes -subj '/CN=pleroma.nixos.test'
+    mkdir -p $out
+    cp key.pem cert.pem $out
+  '';
+
+  # Toot is preventing users from feeding login_cli a password non
+  # interactively. While it makes sense most of the times, it's
+  # preventing us to login in this non-interactive test. This patch
+  # introduce a TOOT_LOGIN_CLI_PASSWORD env variable allowing us to
+  # provide a password to toot login_cli
+  #
+  # If https://github.com/ihabunek/toot/pull/180 gets merged at some
+  # point, feel free to remove this patch.
+  custom-toot = pkgs: pkgs.toot.overrideAttrs(old:{
+    patches = [ (pkgs.fetchpatch {
+      url = "https://github.com/NinjaTrappeur/toot/commit/b4a4c30f41c0cb7e336714c2c4af9bc9bfa0c9f2.patch";
+      sha256 = "sha256-0xxNwjR/fStLjjUUhwzCCfrghRVts+fc+fvVJqVcaFg=";
+    }) ];
+  });
+
+  hosts = nodes: ''
+    ${nodes.pleroma.config.networking.primaryIPAddress} pleroma.nixos.test
+    ${nodes.client.config.networking.primaryIPAddress} client.nixos.test
+  '';
+in import ./make-test-python.nix {
+  name = "pleroma";
+  nodes = {
+    client = { nodes, pkgs, config, ... }: {
+      security.pki.certificateFiles = [ "${tls-cert pkgs}/cert.pem" ];
+      console.keyMap = "fr-bepo";
+      networking.extraHosts = hosts nodes;
+      environment.systemPackages = with pkgs; [
+        (custom-toot pkgs)
+        jq
+        curl
+        (send-toot pkgs)
+      ];
+    };
+    pleroma = { nodes, pkgs, config, ... }: {
+      security.pki.certificateFiles = [ "${tls-cert pkgs}/cert.pem" ];
+      networking.extraHosts = hosts nodes;
+      console.keyMap = "fr-bepo";
+      networking.firewall.enable = false;
+      environment.systemPackages = with pkgs; [
+        postgresql
+        (provision-db pkgs)
+        (provision-secrets pkgs)
+        (provision-user pkgs)
+        sudo
```

Comment:
Sudo should not be here. It usually comes via the module system and is currently a default.